### ISSUE TITLE
Probe plugins by name to avoid cloaking & randomization

### ISSUE THREAD

**Issue by Valve** [Valve/fingerprintjs2#56]
Currently various browser vendors are trying to mitigate browser fingerprinting. 
Firefox does so called "plugin cloaking" - https://wiki.mozilla.org/Fingerprinting#Plugins, i.e. only allows to enumerate the most popular plugins (like Flash and QuickTime), while others are hidden from enumeration. It's still possible to get the availability of the plugin by trying to get it by name:

```js
navigator.plugins["Unity Player"].name // get cloaked plugin by name
"Unity Player"
```

Some other browsers, like Pale Moon, do plugin randomization to always return plugins in a different order:
https://forum.palemoon.org/viewtopic.php?f=26&t=4406

One way to find randomization is to sort the list of plugins. But this will not solve the issue of cloaking.

So I suggest we need to expand the plugin detection code to:

1. Get a list of available plugins calling `navigator.plugins` and then sort this list.
2. Probe additional plugins having a predefined list of plugins. This list should be compiled in this issue.
3. Combine two lists.

For now, I'm including the starting list of plugins to work with:

```
          "AcroPDF.PDF", // Adobe PDF reader 7+
          "Adodb.Stream",
          "AgControl.AgControl", // Silverlight
          "DevalVRXCtrl.DevalVRXCtrl.1",
          "MacromediaFlashPaper.MacromediaFlashPaper",
          "Msxml2.DOMDocument",
          "Msxml2.XMLHTTP",
          "PDF.PdfCtrl", // Adobe PDF reader 6 and earlier, brrr
          "QuickTime.QuickTime", // QuickTime
          "QuickTimeCheckObject.QuickTimeCheck.1",
          "RealPlayer",
          "RealPlayer.RealPlayer(tm) ActiveX Control (32-bit)",
          "RealVideo.RealVideo(tm) ActiveX Control (32-bit)",
          "Scripting.Dictionary",
          "SWCtl.SWCtl", // ShockWave player
          "Shell.UIHelper",
          "ShockwaveFlash.ShockwaveFlash", //flash plugin
          "Skype.Detection",
          "TDCCtl.TDCCtl",
          "WMPlayer.OCX", // Windows media player
          "rmocx.RealPlayer G2 Control",
          "rmocx.RealPlayer G2 Control.1"
```

**Comment by Valve** [comment]
After having thought about it, I can suggest these solutions:

1. We have 2 separate issues: cloaking and randomization. Cloaking will be solved by probing and having a big known list of plugin names.
2. Randomization only exists (to my knowledge) in Pale Moon browser, so we need to make sure to only sort the plugins for this browser only.

@timkolesa suggested to query the plugins twice and sort them if two lists are sorted differently. But I don't think this is possible, because during a page session, the plugin order is consistent, it only changes on the page refresh.

So I'm going to introduce a configuration property, called `sortPluginsFor`, which accepts an array of regexes for which the plugins should be sorted.

By default, this property will be initialized with `/palemoon/i` regex, but an end user of the library will be able to extend the list with more regexes.

**Comment by Valve** [comment]
@timkolesa, I've released the new version (0.5.0), which fixes the palemoon issue.

**Comment by Valve** [comment]
I'm sorry for closing, but we'll be concentrating on stabilizing the library, not on adding new features in the foreseeable future. Closing as a non-feature.
